### PR TITLE
SONAR-7986 WS settings/reset check scope and qualifier

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SetAction.java [hunk 1]
```diff
@@ -79,16 +82,18 @@
   private final UserSession userSession;
   private final SettingsUpdater settingsUpdater;
   private final SettingsChangeNotifier settingsChangeNotifier;
+  private final SettingValidator settingValidator;
 
   public SetAction(PropertyDefinitions propertyDefinitions, I18n i18n, DbClient dbClient, ComponentFinder componentFinder, UserSession userSession,
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SetAction.java]: MAJOR Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidator.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  public SettingValidation validateQualifier() {
+    return data -> {
+      String qualifier = data.component == null ? "" : data.component.qualifier();
+      checkRequest(data.component == null || data.definition == null || data.definition.qualifiers().contains(data.component.qualifier()),
+        "Setting '%s' cannot be set on a %s", data.key, i18n.message(Locale.ENGLISH, "qualifier." + qualifier, null));
+    };
+  }
+
+  @FunctionalInterface
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingValidator.java]: Interesting ! As I'm not used to this new API, it takes will some time to understand this class, but now I feel it's good, well done !

**Comment by julienlancelot** [comment]
LGTM, I'm merging it to master

**Pull request merged**